### PR TITLE
Change LogAcceptCategory to use uint32_t rather than sets of strings.

### PR DESCRIPTION
**Pull request by gmaxwell**
This changes the logging categories to boolean flags instead of strings.

This simplifies the acceptance testing by avoiding accessing a scoped
 static thread local pointer to a thread local set of strings.  It
 eliminates the only use of boost::thread_specific_ptr outside of
 lockorder debugging.

This change allows log entries to be directed to multiple categories
 and makes it easy to change the logging flags at runtime (e.g. via
 an RPC, though that isn't done by this commit.)

It also eliminates the fDebug global.

Configuration of unknown logging categories now produces a warning.

### CODE REVIEW

**Comment by gmaxwell** [comment]
If we ever need more than 32 categories (there are 19 now) this could be trivially changed to use a uint64_t.

File: src/net_processing.cpp [hunk 1]
```diff
@@ -1432,11 +1432,10 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
...
         }
 
-        if (fDebug || (vInv.size() != 1))
-            LogPrint("net", "received getdata (%u invsz) peer=%d\n", vInv.size(), pfrom->id);
+        LogPrint(LOG_NET, "received getdata (%u invsz) peer=%d\n", vInv.size(), pfrom->id);
 
-        if ((fDebug && vInv.size() > 0) || (vInv.size() == 1))
-            LogPrint("net", "received getdata for: %s peer=%d\n", vInv[0].ToString(), pfrom->id);
+        if (vInv.size() > 0)
+            LogPrint(LOG_NET, "received getdata for: %s peer=%d\n", vInv[0].ToString(), pfrom->id);
```
Comment (gmaxwell) [inline src/net_processing.cpp]: The prior code above struck me as somewhat brain-damaged (LogPrint() does nothing when fDebug isn't true), so this chunk could use some more review attention in case I was missing something.

File: src/util.h [hunk 2]
```diff
@@ -69,15 +70,50 @@ inline std::string _(const char* psz)
 void SetupEnvironment();
 bool SetupNetworking();
 
+enum LogFlags : uint32_t {
```
Comment (rebroad) [inline src/util.h]: idea/nit: Would be easier to maintain if it was possible to define LogFlags and CLogCategoryDesc in the same section...

**Comment by TheBlueMatt** [comment]
Concept ACK

**Comment by rebroad** [comment]
ACK

**Comment by MarcoFalke** [comment]
utACK e348647315b619e6a1d8b614554abb0747047696

**Comment by paveljanik** [comment]
Needs rebase.

**Comment by paveljanik** [comment]
@gmaxwell Please reopen this...

**Comment by jonasschnelli** [comment]
Concept ACK, seems to be a great PR with serval acks. Needs rebase and reopen.

**Comment by jtimon** [comment]
Concept ACK

**Comment by jnewbery** [comment]
Definite concept ACK. Makes #10123 trivial.

**Comment by laanwj** [comment]
Concept ACK.

File: src/util.h [hunk 3]
```diff
@@ -69,15 +70,50 @@ inline std::string _(const char* psz)
...
+    LOG_NONE        = 0,
+    LOG_NET         = (1 <<  0),
```
Comment (laanwj) [inline src/util.h]: Instead of the LOG_ prefix, I'd propose to use a scoped enum. E.g. `LogFlags::NONE` `LogFlags::NET` etc.

Comment (gmaxwell) [inline src/util.h]: Oh darn, didn't see this until I'd already done the rebase. Uh, I'll look into it and put it as a commit on top for squashing.  (presumably it'll be a more or less trivial change that could be reviewed separately)

**Review by JeremyRubin — COMMENTED** [review]
Concept ACK!

I think it would be nice if there were a semi-convenient way to add new categories without triggering a big recompile though...

File: src/util.h (hunk 3 above)
Comment (gmaxwell) [inline src/util.h]: 19:32 < gmaxwell> wumpus: uh. so boolean operations don't appear to exist for scoped enums. so ... I don't think I can use them without 
                  defining a set of boolean operations for them. 
19:32 < sipa> gmaxwell: eh?
19:33 < sipa> ah, scoped enums don't have an implicit conversion to int
19:34 < sipa> you can still explicitly convert them, though
19:34 < sipa> but enums are generally a bad approach for bitfields, as they're intended to be exhaustive
19:34 < gmaxwell> sipa: wumpus suggested that I change the log print thing to use a scoped enum.  The original thing that inspired this 
                  change is that rebroad wanted to do some horrible hack to do the logical equivilent of LogPrint(LOG_NET | LOG_MEMPOOL, 
                  ...). (and of course the matching also uses it as a bitfield.)
19:35 < sipa> right, a normal enum (which behaves much more like an int) may be more appropriate
19:35 < gmaxwell> well thats what I implemented.
19:37 < sipa> awesöme
19:37 < gmaxwell> in any case. I'm happy to change it to whatever. someone who has an opinion needs to tell me what to do. I think I only 
                  need a couple casts to make the enum class stuff go, but it would be ugly to do "LogPrint(LOG_NET | LOG_MEMPOOL," later 
                  unless we provide | for the type.

File: src/util.h [hunk 4]
```diff
@@ -56,6 +55,8 @@ extern CTranslationInterface translationInterface;
 extern const char * const BITCOIN_CONF_FILENAME;
 extern const char * const BITCOIN_PID_FILENAME;
 
+extern std::atomic<uint32_t> logCategories;
```
Comment (gmaxwell) [inline src/util.h]: I expect the very next logical change will be to add an RPC to turn on and off logging categories. Since all the reads are order relaxed I expect it to emit the same code as a non-atomic read (at least on x86).

Comment (theuni) [inline src/util.h]: I've wanted this for so long! Consider this the next/2 logical change.

File: src/util.h (hunk 3 above)
Comment (laanwj) [inline src/util.h]: It's just that the LOG_ prefix is very common, so it's easy to get namespace collisions. Scoped enums avoid that, the only way any of their members can collide is when the name of the type conflicts, which would be an error in the first place.
Then again if this introduces extra C++ casting uglyness in all usages let's leave it like this...

File: src/util.h [hunk 5]
```diff
@@ -69,15 +70,50 @@ inline std::string _(const char* psz)
...
+
 /** Send a string to the log output */
 int LogPrintStr(const std::string &str);
 
-#define LogPrintf(...) LogPrint(NULL, __VA_ARGS__)
+#define LogPrintf(...) LogPrintStr(tfm::format(__VA_ARGS__))
 
 template<typename... Args>
-static inline int LogPrint(const char* category, const char* fmt, const Args&... args)
+static inline int LogPrint(uint32_t category, const char* fmt, const Args&... args)
```
Comment (laanwj) [inline src/util.h]: Yes, that would add some type safety. It would make it more of a hassle to submit a message to more categories at once but that isn't supported now either.

**Comment by JeremyRubin** [comment]
Suggestion on how to make the category more extensible without triggering big recompiles:
```c++
#define LogPrint(cat, ...) _LogPrint(cat, #cat, __VA_ARGS__)
static inline int _LogPrint(uint32_t category, const char* category_str, const char* fmt, const Args&... args) {
    // ...
    if (category == LOG_USE_IDENT_NAME)
        // use the macro
    // ...
}
```

then from `somefile.cpp`

```c++
static const uint32_t LOG_SOMEFILE = LOG_USE_IDENT_NAME;
LogPrint(LOG_SOMEFILE, "...");

```

File: src/util.h (hunk 3 above)
Comment (gmaxwell) [inline src/util.h]: I could make it BCLOG or something like that, you're right about namespace collisions, my initial sed through to rename the LOG_ accidentally renamed some things it shouldn't.

Comment (sipa) [inline src/util.h]: Another approach would be wrapping it in a C++ namespace (for example:
```c++
namespace bclog {
enum flags {
  NET,
  MEMPOOL,
  ...
};
}
```

at which point you can use `bclog::NET | bclog::MEMPOOL`.

Comment (jnewbery) [inline src/util.h]: BITCOIN_LOG_* or similar is unlikely to collide with anything.

File: src/util.cpp [hunk 6]
```diff
@@ -227,36 +228,69 @@ void OpenDebugLog()
...
+    {LOG_CMPCTBLOCK, "cmpctblock"},
+    {LOG_RAND, "rand"},
+    {LOG_PRUNE, "prune"},
+    {LOG_PROXY, "proxy"},
+    {LOG_MEMPOOLREJ, "mempoolrej"},
+    {LOG_LIBEVENT, "libevent"},
+    {LOG_COINDB, "coindb"},
+    {LOG_QT, "qt"},
+    {LOG_LEVELDB, "leveldb"},
+    {LOG_ALL, "1"},
```
Comment (jnewbery) [inline src/util.cpp]: Can you add an "all" alias here, ie `-debug=all` is equivalent to `-debug=1`

File: src/init.cpp [hunk 7]
```diff
@@ -908,12 +905,23 @@ bool AppInitParameterInteraction()
 
     // ********************************************************* Step 3: parameter-to-internal-flags
 
-    fDebug = mapMultiArgs.count("-debug");
+    bool fDebug = mapMultiArgs.count("-debug") > 0;
```
Comment (jnewbery) [inline src/init.cpp]: Now that you've removed the global fDebug bool, you can go ahead and remove it entirely and test the conditions directly (rather than make the variable local), ie:

`if (mapMultiArgs.count("-debug") > 0) {`

and further below:

`if !(GetBoolArg("-nodebug", false) || find(categories.begin(), categories.end(), std::string("0")) != categories.end()) {`

File: src/util.h [hunk 8]
```diff
@@ -68,30 +69,55 @@ inline std::string _(const char* psz)
...
-    } \
-    LogPrintStr(_log_msg_); \
-} while(0)
-
-#define LogPrint(category, ...) do { \
-    if (LogAcceptCategory((category))) { \
-        LogPrintf(__VA_ARGS__); \
-    } \
-} while(0)
+#define LogPrintf(...) LogPrintStr(tfm::format(__VA_ARGS__))
```
Comment (jnewbery) [inline src/util.h]: Why have you removed the try/except handling in this macro? (I'm not saying it's wrong, I just don't understand why it was necessary before but not now).

Comment (jnewbery) [inline src/util.h]: EDIT: I think this is a bad rebase that's reverted #9963

File: src/tinyformat.h [hunk 9]
```diff
@@ -167,7 +167,7 @@ namespace tinyformat {
 class format_error: public std::runtime_error
 {
 public:
-    format_error(const std::string &what): std::runtime_error(what) {
+    format_error(const std::string &fmtwhat): std::runtime_error(fmtwhat) {
```
Comment (jnewbery) [inline src/tinyformat.h]: bad rebase?

**Review by jnewbery — COMMENTED** [review]
It looks to me like there are a couple of rebase errors here. Other than that a few nits but looks good. I've tested ab53beaf9c0d8f923f24a50b373968a8c01760d4 manually and against the extended test suite and it works.

One thing to note about the future logging RPC work: I think libevent and leveldb logging levels are set at start up, so they won't be modifiable through the RPC without some additional work. Not a problem for this PR, but something to note since it'd be easy to believe that just updating logCategories would be enough. Perhaps worth putting a comment next to the logCategories variable warning about this?

File: src/tinyformat.h (hunk 9 above)
Comment (gmaxwell) [inline src/tinyformat.h]: No. silencing an obnoxious shadowing warning that was making it impossible to see where other issues were.

File: src/init.cpp (hunk 7 above)
Comment (gmaxwell) [inline src/init.cpp]: Great point.

File: src/util.h (hunk 8 above)
Comment (gmaxwell) [inline src/util.h]: Yep. Fixing.

File: src/util.cpp (hunk 6 above)
Comment (gmaxwell) [inline src/util.cpp]: Yup.

**Comment by gmaxwell** [comment]
I think I addressed all the comments.

File: src/tinyformat.h (hunk 9 above)
Comment (jnewbery) [inline src/tinyformat.h]: My only concern is that we occasionally pull updates to tinyformat.cpp from upstream (eg #8274), so any trivial changes like this add to the maintainer's burden when merging. Not sure if @laanwj has an opinion.

Comment (laanwj) [inline src/tinyformat.h]: Can we just disable the shadowing warnings please, instead of peppering unrelated variable renames in commits?

Comment (laanwj) [inline src/tinyformat.h]: There you go: #10136

**Comment by laanwj** [comment]
> Suggestion on how to make the category more extensible without triggering big recompiles:

IMO we shouldn't worry about recompiles when the set of categories changes. That's a rare occurrence, and the way headers depend on each other almost every change already results in a (near) full recompile.

**Comment by laanwj** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/9424/commits/21d94e712ddd7859bfaf02282eaa38cb18ec7e1b apart from the variable name change in `tinyformat.h`

**Comment by gmaxwell** [comment]
(Trivally) rebased, squashed, and removed the tinyformat change.

File: src/util.h (hunk 4 above)
Comment (laanwj) [inline src/util.h]: Yes being able to change the debug flags at runtime (even if not the libevent and leveldb ones which are processed at startup) would be nice.

**Comment by laanwj** [comment]
Going to merge this; a change like this will need rebase after every single thing that gets merged, and @gmaxwell addressed all the nits. Further improvements can be done in later PRs.
